src/tactic/induction.lean

```diff
@@ -0,0 +1,1290 @@
+  rhs_eq_lhs ← succeeds $ unify lhs rhs,
+  if rhs_eq_lhs
+    then pure ((const `eq.refl [u]) type lhs, none)
+    else do
+      pure (h, some h)
+| h@(local_const uname ppname binfo
+    T@(app (app (app (app (const `heq [u]) lhs_type) lhs) rhs_type) rhs)) := do
+  lhs_type_eq_rhs_type ← succeeds $ is_def_eq lhs_type rhs_type,
+  if ¬ lhs_type_eq_rhs_type
+    then do
+      pure (h, some h)
+    else do
+      lhs_eq_rhs ← succeeds $ unify lhs rhs,
+      if lhs_eq_rhs
+        then pure ((const `heq.refl [u]) lhs_type lhs, none)
+        else do
+          c ← mk_local' ppname binfo $ (const `eq [u]) lhs_type lhs rhs,
+          let arg := (const `heq_of_eq [u]) lhs_type lhs rhs c,
+          pure (arg, some c)
+| (local_const _ _ _ T) := fail!
+  "process_index_equation: expected a homogeneous or heterogeneous equation, but got:\n{T}"
+| e := fail!
+  "process_index_equation: expected a local constant, but got:\n{e}"
+
+/--
+`assign_local_to_unassigned_mvar mv pp_name binfo`, where `mv` is a
+metavariable, acts as follows:
+
+- If `mv` is assigned, it is not changed and the tactic returns `none`.
+- If `mv` is not assigned, it is assigned a fresh local constant with
+  the type of `mv`, pretty name `pp_name` and binder info `binfo`. This local
+  constant is returned.
+-/
+meta def assign_local_to_unassigned_mvar (mv : expr) (pp_name : name)
+  (binfo : binder_info) : tactic (option expr) := do
+  ff ← is_assigned mv | pure none,
+  type ← infer_type mv,
+  c ← mk_local' pp_name binfo type,
+  unify mv c,
+  pure c
+
+/--
+Apply `assign_local_to_unassigned_mvar` to a list of metavariables. Returns the
+newly created local constants.
+-/
+meta def assign_locals_to_unassigned_mvars
+  (mvars : list (expr × name × binder_info)) : tactic (list expr) :=
+mvars.mmap_filter $ λ ⟨mv, pp_name, binfo⟩,
+  assign_local_to_unassigned_mvar mv pp_name binfo
+
+/--
+Simplify an induction hypothesis.
+
+Input: a local constant
+```
+ih : ∀ (x₁ : T₁) ... (xₙ : Tₙ) (eq₁ : y₁ = z₁) ... (eqₘ : yₘ = zₘ), P
+```
+where `n = num_generalized` and `m = num_index_vars`. The `xᵢ` are hypotheses
+that we generalised over before performing induction. The `eqᵢ` are index
+equations.
+
+Output: a new local constant
+```
+ih' : ∀ (x'₁ : T'₁) ... (x'ₖ : T'ₖ) (eq'₁ : y'₁ = z'₁) ... (eq'ₗ : y'ₗ = z'ₗ), P'
+```
+This new induction hypothesis is derived from `ih` by removing those `eqᵢ` whose
+left- and right-hand sides can be unified. This unification may also determine
+some of the `xᵢ`. The `x'ᵢ` and `eq'ᵢ` are those `xᵢ` and `eqᵢ` that were not
+removed by this process.
+
+Some of the `eqᵢ` may be heterogeneous: `eqᵢ : yᵢ == zᵢ`. In this case, we
+proceed as follows:
+
+- If `yᵢ` and `zᵢ` are defeq, then `eqᵢ` is removed.
+- If `yᵢ` and `zᵢ` are not defeq but their types are, then `eqᵢ` is replaced by
+  `eq'ᵢ : x = y`.
+- Otherwise `eqᵢ` remains unchanged.
+-/
+meta def simplify_ih (num_generalized : ℕ) (num_index_vars : ℕ) (ih : expr) :
```

Comment:
Is there a reason you don't call `unify_equations` to simplify the induction hypothesis? That seems useful, so that you can also simplify equations like `succ n = succ m` and `0 = succ n`.

Reply:
I believe your examples are covered by the current algorithm: for an IH `forall x y, x.succ = y.succ -> P x y`, we get `forall y, P y y`. I use Lean's builtin unification with metavariables mostly because I would otherwise have to majorly refactor `unify_equations`.

Reply:
Ah, I see. Since `succ ?n` and `succ ?m` can be unified, `process_index_equation` will be able to simplify it. 
Where do you deal with the case `0 = succ n`? `process_index_equation` doesn't do anything with an equation of that form, correct?

And `unify_equations` has more bells and whistles if I understand correctly, like simplifying `t.succ = s.succ` to `t = s` when `t` and `s` are not variables, or simplifying `n = succ (succ n)` to `false`.

It's unfortunate that you have two different methods to deal with (essentially?) the same task, namely to simplify equations generated by `induction'`, but I understand that it might be tough to integrate the two methods if that was not planned from the start.

Reply:
You're right that Lean's builtin unification is more limited in the cases you mention. In particular, it doesn't distinguish between 'unification failed but the terms could be defeq' and 'unification failed because the terms are definitely not defeq', so we can't clear a hyp which assumes e.g. `0 = 1`. (On the other hand, Lean's unification supports a limited form of higher-order unification, which `unify_equations` doesn't do at all. Don't think that's relevant for this application though.)

In principle, it wouldn't be a problem to apply the `unify_equations` algorithm here as well. It would just need a complete refactoring of `unify_equations`, so I'd like to leave that as a TODO for now. I'll add a note to that effect.